### ISSUE TITLE
New-HVPool & Nics

### ISSUE THREAD

**Issue by breezytm** [vmware/PowerCLI-Example-Scripts#376]
Hi, 

I am writing a script to automate some pools creation since we are migrating from linked clones to instant clones and I have been facing an issue when it comes to the -Nics parameter. I know it's optional but due to VLANs and DCHP restrictions I would like to assign each pool their own network(s). Anyway someone can provide me a sample on how to use the parameter using a port group label as string. 

-Nics <DesktopNetworkInterfaceCardSettings[]>
desktopSpec.automatedDesktopSpec.virtualCenterProvisioningSettings.virtualCenterNetworkingSettings.nics

**Comment by mattivoxy** [comment]
@breezytm 

Thanks for pressing on this. Couple things:
- Your script is showing as password protected, so I can't grab and take a look
- I get a similar error, but the calling error calls out "Desktop_Update" for me. Oddly enough, this *sometimes* works. If I paste it all in manually into a PS shell, it will work for me. Very strange.

**Comment by mattivoxy** [comment]
Of note: this will only work once the pool is created and published; the "snapshot" property doesn't exist until the parent VM is published. Since you're trying to do this @ creation, it will certainly fail as the snapshot doesn't exist yet. Here's the line where it's breaking, most likely:

$NetworkInterfaceCardInfo = $NetworkInterfaceCardService.NetworkInterfaceCard_ListBySnapshot($services, **$PoolInfo.AutomatedDesktopData.VirtualCenterProvisioningSettings.VirtualCenterProvisioningData.Snapshot**)

So basically, we need to wait for the pool to completely provision/publish, and then we can make the change. Now, how to programatically sniff that out, or just setup a ~5m timer (ick)?

**Comment by breezytm** [comment]
@mattivoxy 

The password to the script is VMware! 
I password protected it because it is in early stage and I need to do a lot of clean up. I didn't want to show it to the world yet.

> So basically, we need to wait for the pool to completely provision/publish, and then we can make the change. Now, how to programatically sniff that out, or just setup a ~5m timer (ick)?

So far I've seen only two people used it and they both were updating a pool. I don't understand why would the cmdlet has a parameter that can't be used during pool creation.

**Comment by mattivoxy** [comment]
Is it necessary to reference $PoolInfo.AutomatedDesktopData.VirtualCenterProvisioningSettings.VirtualCenterProvisioningData.Snapshot as part of this operation? That is what doesn't exist until the pool is published, which takes some time obviously. If that part is unnecessary maybe we can get around that.

**Comment by freythman** [comment]
@breezytm Taking a look at your script, the `New-HVPool` command is using `-nics`, which to be clear, **is not yet implemented.** 

See this line for what I mean by that.
https://github.com/vmware/PowerCLI-Example-Scripts/blob/d64e0fbdc09309a3c7261d0faa9371695f64b9ef/Modules/VMware.Hv.Helper/VMware.HV.Helper.psm1#L4462 

That being said, as a workaround you could create the pool as you do today, without specifying the `-nic` parameter. Make sure that `-enableProvisioning` is set to `$false` to prevent the new machines from being created.

Then follow my example of using `-key` and `-value` along with `Set-HVPool` (including the fixes that you reference in https://github.com/vmware/PowerCLI-Example-Scripts/issues/376#issuecomment-678787458).

Then in the next step, enable provisioning by running `Set-HVPool -Key 'automatedDesktopData.virtualCenterProvisioningSettings.enableProvisioning' -Value $true`

I realize this isn't an ideal workflow, but it's all I can offer at the moment, until I, or someone else, has time to flesh out the `-nics` parameter to both `New-HVPool` and `Set-HVPool`.

**Comment by freythman** [comment]
> Is it necessary to reference $PoolInfo.AutomatedDesktopData.VirtualCenterProvisioningSettings.VirtualCenterProvisioningData.Snapshot as part of this operation? That is what doesn't exist until the pool is published, which takes some time obviously. If that part is unnecessary maybe we can get around that.

@mattivoxy This only needs to be referenced for existing pools. It was done this way to ensure that when editing an existing pool, we reference the same `BaseImageSnapshotId` that is in use for the pool.

For new pools, if you were to manually build the `DesktopSpec`, then you could obtain the `BaseImageSnapshotId` using `BaseImageSnapshot_List()` from the `BaseImageSnapshot` service. This method does require inputting a `BaseImageVmId` which can be obtained using `BaseImageVm_List()`, an example of which can be seen here:  https://github.com/vmware/PowerCLI-Example-Scripts/blob/d64e0fbdc09309a3c7261d0faa9371695f64b9ef/Modules/VMware.Hv.Helper/VMware.HV.Helper.psm1#L4959

More information on the `BaseImageSnapshot` service: https://vdc-download.vmware.com/vmwb-repository/dcr-public/93e1f7c4-fae6-4e10-b3be-7c73720b4589/2f59c85d-2770-45cd-bd9d-3a5cc747f3a3/vdi.utils.virtualcenter.BaseImageSnapshot.html

**Comment by breezytm** [comment]
@freythman I did exactly what you mentioned and am still getting the error. 

I removed the `-nics` parameter as well as setting `-enableprovisioning $false` but I still get the same error.


No error is retured with -WhatIf so i don't know where the problem is coming from at this point and tomorrow is the day. I am running out of time. 

```
What if: Performing the operation "New-HVPool" on target "TESTVM".

Base                  : VMware.Hv.DesktopBase
DesktopSettings       : VMware.Hv.DesktopSettings
Type                  : AUTOMATED
AutomatedDesktopSpec  : VMware.Hv.DesktopAutomatedDesktopSpec
ManualDesktopSpec     : 
RdsDesktopSpec        : 
GlobalEntitlementData : 
```

**Comment by freythman** [comment]
The function has a debug mode, that is currently disabled. Can you uncomment this code block and re-run your script? That will output a .json file with the full output of the `DesktopSpec.` 

Optionally, you can also enable verbose output in PowerShell, which may (or may not) offer more useful information. 

`$VerbosePreference = 'Continue'`

The error you referenced above means that something with the spec is incorrect. At a glance, you can remove the following parameters to simplify troubleshooting this error since they are set as `$null` in your .csv file.

```
-SpecificNames
-PowerOffScriptName
-PowerOffScriptParameters
-PostSynchronizationScriptName
-ConnectionServerRestrictions
-GlobalEntitlement
```

**Comment by breezytm** [comment]
@freythman 

Understood. As I troubleshoot further, I see there may be a potential issues with my csv containing the word $null. It may not like that. So far the items you mentioned above to remove where all set to $null

**Comment by breezytm** [comment]
@freythman 

Here's an update. I was able to identify the issue with the pool creation. It turns out to be three things. 

1. `-ConnectionServerRestrictions` must have a value. It does not accept neither null nor an empty string. I am still trying to figure out how to code around that limitation. 
2. `-Datastores, -StorageOvercommit, `must be in this format `Datastore1,Datastore2,Datastore3 and UNBOUNDED,UNBOUNDED,UNBOUNDED` 

However, the `-nics` is still an issue. If you can take a look at the code you send me with the changes I made and see if you can pinpoint the issue. 

```
    #Initiated service and objects
    $services = $HVServer.ExtensionData
    $DataCenterService = New-Object VMware.Hv.DatacenterService
    $HostOrClusterService = New-Object VMware.Hv.HostOrClusterService
    $NetworkLabelService = New-Object VMware.Hv.NetworkLabelService
    $NetworkInterfaceCardService = New-Object VMware.Hv.NetworkInterfaceCardService


    $VirtualCenterInfo = Get-HVvCenterServer -Name $Pool.Vcenter
    $DataCenterInfo = $DataCenterService.Datacenter_List($services, $VirtualCenterInfo.Id) | Where-Object { $_.Base.Name -eq $Pool.datacenter }
    $HostOrClusterList = $HostOrClusterService.HostOrCluster_GetHostOrClusterTree($services, $DataCenterInfo.Id)
    $ClusterInfo = $HostOrClusterList.TreeContainer.Children.Info | Where-Object { $_.Name -eq $Pool.HostOrCluster }


    $NetworkLabelList = $NetworkLabelService.NetworkLabel_ListByHostOrCluster($services, $ClusterInfo.Id)
    $NetworkLabelInfo = $NetworkLabelList | Where-Object {$_.Data.Name -eq $Pool.Nics }


    # Build the spec
    $DesktopNetworklabelAssignmentSpec = New-Object VMware.Hv.DesktopNetworkLabelAssignmentSpec
    $DesktopNetworklabelAssignmentSpec.enabled = $true
    $DesktopNetworklabelAssignmentSpec.networkLabel = $NetworkLabelInfo.Id
    #Change these according to https://vdc-download.vmware.com/vmwb-repository/dcr-public/93e1f7c4-fae6-4e10-b3be-7c73720b4589/2f59c85d-2770-45cd-bd9d-3a5cc747f3a3/vdi.resources.Desktop.NetworkLabelAssignmentSpec.html
    $DesktopNetworklabelAssignmentSpec.maxLabelType = $Pool.maxLabelType
    $DesktopNetworklabelAssignmentSpec.maxLabel = $Pool.maxLabel

    $DesktopNetworkInterfaceCardSettings = New-Object VMware.Hv.DesktopNetworkInterfaceCardSettings
    $DesktopNetworkInterfaceCardSettings.Nic = $NetworkInterfaceCardInfo.Id
    $DesktopNetworkInterfaceCardSettings.NetworkLabelAssignmentSpecs = $DesktopNetworklabelAssignmentSpec

    [VMware.Hv.DesktopNetworkInterfaceCardSettings[]]$niclist = @()
    $niclist += $DesktopNetworkInterfaceCardSettings

    #assign nics to existing pool
    Set-HVPool -PoolName $Pool.PoolName -Key "automatedDesktopData.virtualCenterProvisioningSettings.virtualCenterNetworkingSettings.nics" -Value $niclist
```
Error 

```
Updating the Pool:  TESTVM
Exception calling "Desktop_Update" with "3" argument(s): "There is an error in the XML document."
At C:\Program Files\WindowsPowerShell\Modules\VMware.Hv.Helper\VMware.HV.Helper.psm1:6362 char:8
+        $desktop_helper.Desktop_Update($services,$item,$updates)
+        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : InvalidOperationException
```
